File: src/geoseal_operator_space.py
"""GeoSeal Operator System-Space Model (Python reference implementation).

Extends GeoSeal's geographic ring model to cover the operator's position in
**system topology space** — not physical coordinates, but the structural
location of the operator: which access plane they arrived through (web /
terminal / app / API) and whether they are authenticated.

The access plane + auth state together determine:
  - A file-system topology (what paths are reachable and writable)
  - A governance ring (core / outer / restricted / blocked)
  - A deterministic space ID analogous to the geographic geoid
  - Governance flags that surface cross-plane claims and auth anomalies

Why this matters:
  A web-anonymous operator claiming /home/user/documents is structurally
  impossible — their sandbox grants no native FS at all. The governance layer
  should catch the claim before it reaches the FS.  Similarly, a terminal+sudo
  operator carries elevated risk that warrants its own flag even at high trust.
"""

from __future__ import annotations

import hashlib
import re
from dataclasses import dataclass, field
from typing import List, Literal, Optional

# ---------------------------------------------------------------------------
# Types / enumerations
# ---------------------------------------------------------------------------

AccessPlane = Literal["web", "terminal", "app", "api"]
AuthState = Literal["authenticated", "anonymous", "service_account", "sudo"]
SandboxLevel = Literal["none", "partial", "full"]
OperatorRing = Literal["core", "outer", "restricted", "blocked"]
GovernanceTier = Literal["ALLOW", "QUARANTINE", "ESCALATE", "DENY"]

GOVERNANCE_FLAGS = {
    "ELEVATED_TERMINAL",
    "UNAUTHENTICATED_WEB",
    "UNAUTHENTICATED_API",
    "CROSS_PLANE_CLAIM",
    "SESSION_MISSING_LOGIN_TIME",
    "TEMP_FS_ONLY",
    "SERVICE_ACCOUNT_ELEVATED",
}


# ---------------------------------------------------------------------------
# Data models
# ---------------------------------------------------------------------------


@dataclass(frozen=True)
class FsTopology:
    """File-system access topology for an operator's plane + auth combination."""

    sandbox_level: SandboxLevel
    accessible_roots: List[str]
    writable_paths: List[str]
    persists_across_sessions: bool
    temp_only: bool
    mount_points: List[str]


@dataclass(frozen=True)
class OperatorSpaceInput:
    """Input to the operator space evaluator."""

    access_plane: AccessPlane
    auth_state: AuthState
    session_fingerprint: Optional[str] = None
    login_time_ms: Optional[int] = None
    claimed_paths: List[str] = field(default_factory=list)


@dataclass(frozen=True)
class OperatorSpaceDecision:
    """Governance decision for an operator's system-space position."""

    ring: OperatorRing
    trust_score: float
    space_id: str
    fs_topology: FsTopology
    governance_flags: List[str]
    status: str
    reason: str


# ---------------------------------------------------------------------------
# FS topology derivation
# ---------------------------------------------------------------------------

_FS_TOPOLOGY: dict[tuple[AccessPlane, AuthState], FsTopology] = {
    # --- web ---
    ("web", "anonymous"): FsTopology(
        sandbox_level="full",
        accessible_roots=[],
        writable_paths=[],
        persists_across_sessions=False,
        temp_only=True,
        mount_points=[],
    ),
    ("web", "authenticated"): FsTopology(
        sandbox_level="full",
        accessible_roots=[],
        writable_paths=[],
        persists_across_sessions=True,
        temp_only=False,
        mount_points=[],
    ),
    ("web", "service_account"): FsTopology(
        sandbox_level="full",
        accessible_roots=[],
        writable_paths=[],
        persists_across_sessions=True,
        temp_only=False,
        mount_points=[],
    ),
    ("web", "sudo"): FsTopology(
        sandbox_level="full",
        accessible_roots=[],
        writable_paths=[],
        persists_across_sessions=True,
        temp_only=False,
        mount_points=[],
    ),
    # --- terminal ---
    ("terminal", "sudo"): FsTopology(
        sandbox_level="none",
        accessible_roots=["/", "C:\\"],
        writable_paths=["/", "C:\\"],
        persists_across_sessions=True,
        temp_only=False,
        mount_points=["/", "/proc", "/sys", "/dev", "C:\\", "D:\\"],
    ),
    ("terminal", "authenticated"): FsTopology(
        sandbox_level="none",
        accessible_roots=["~", "/home", "/var", "/tmp", "C:\\Users", "C:\\ProgramData"],
        writable_paths=["~", "/tmp", "C:\\Users\\{user}", "C:\\Temp"],
        persists_across_sessions=True,
        temp_only=False,
        mount_points=["/", "C:\\", "D:\\"],
    ),
    ("terminal", "service_account"): FsTopology(
        sandbox_level="none",
        accessible_roots=["~", "/home", "/var", "/tmp", "C:\\Users", "C:\\ProgramData"],
        writable_paths=["~", "/tmp", "C:\\Users\\{user}", "C:\\Temp"],
        persists_across_sessions=True,
        temp_only=False,
        mount_points=["/", "C:\\", "D:\\"],
    ),
    ("terminal", "anonymous"): FsTopology(
        sandbox_level="partial",
        accessible_roots=["/tmp", "C:\\Temp"],
        writable_paths=["/tmp", "C:\\Temp"],
        persists_across_sessions=False,
        temp_only=True,
        mount_points=[],
    ),
    # --- app ---
    ("app", "authenticated"): FsTopology(
        sandbox_level="partial",
        accessible_roots=["~/AppData", "~/.local/share", "~/Library/Application Support"],
        writable_paths=["~/AppData/{app}", "~/.local/share/{app}"],
        persists_across_sessions=True,
        temp_only=False,
        mount_points=[],
    ),
    ("app", "service_account"): FsTopology(
        sandbox_level="partial",
        accessible_roots=["~/AppData", "~/.local/share", "~/Library/Application Support"],
        writable_paths=["~/AppData/{app}", "~/.local/share/{app}"],
        persists_across_sessions=True,
        temp_only=False,
        mount_points=[],
    ),
    ("app", "anonymous"): FsTopology(
        sandbox_level="partial",
        accessible_roots=[],
        writable_paths=[],
        persists_across_sessions=False,
        temp_only=True,
        mount_points=[],
    ),
    ("app", "sudo"): FsTopology(
        sandbox_level="partial",
        accessible_roots=["~/AppData", "~/.local/share", "~/Library/Application Support"],
        writable_paths=["~/AppData/{app}", "~/.local/share/{app}"],
        persists_across_sessions=True,
        temp_only=False,
        mount_points=[],
    ),
    # --- api ---
    ("api", "service_account"): FsTopology(
        sandbox_level="partial",
        accessible_roots=[],
        writable_paths=[],
        persists_across_sessions=False,
        temp_only=True,
        mount_points=[],
    ),
    ("api", "authenticated"): FsTopology(
        sandbox_level="partial",
        accessible_roots=[],
        writable_paths=[],
        persists_across_sessions=False,
        temp_only=True,
        mount_points=[],
    ),
    ("api", "anonymous"): FsTopology(
        sandbox_level="full",
        accessible_roots=[],
        writable_paths=[],
        persists_across_sessions=False,
        temp_only=True,
        mount_points=[],
    ),
    ("api", "sudo"): FsTopology(
        sandbox_level="partial",
        accessible_roots=[],
        writable_paths=[],
        persists_across_sessions=False,
        temp_only=True,
        mount_points=[],
    ),
}


def derive_fs_topology(plane: AccessPlane, auth: AuthState) -> FsTopology:
    """Return the FS topology for a given plane + auth combination."""
    return _FS_TOPOLOGY[(plane, auth)]


# ---------------------------------------------------------------------------
# Ring + trust lookup
# ---------------------------------------------------------------------------

_RING_TABLE: dict[tuple[AccessPlane, AuthState], tuple[OperatorRing, float]] = {
    ("terminal", "sudo"): ("core", 0.95),
    ("terminal", "authenticated"): ("core", 0.85),
    ("terminal", "service_account"): ("core", 0.80),
    ("terminal", "anonymous"): ("restricted", 0.35),
    ("app", "authenticated"): ("outer", 0.70),
    ("app", "service_account"): ("outer", 0.65),
    ("app", "anonymous"): ("restricted", 0.25),
    ("app", "sudo"): ("outer", 0.70),
    ("api", "service_account"): ("core", 0.80),
    ("api", "authenticated"): ("outer", 0.70),
    ("api", "anonymous"): ("blocked", 0.05),
    ("api", "sudo"): ("outer", 0.70),
    ("web", "authenticated"): ("outer", 0.60),
    ("web", "anonymous"): ("blocked", 0.05),
    ("web", "service_account"): ("outer", 0.55),
    ("web", "sudo"): ("outer", 0.60),
}


# ---------------------------------------------------------------------------
# Space ID
# ---------------------------------------------------------------------------

_WEB_NATIVE_PATH = re.compile(r"^[/\\]|^[A-Za-z]:\\")
_URL = re.compile(r"^https?://", re.IGNORECASE)
_UNC = re.compile(r"^\\\\")


def _derive_space_id(plane: AccessPlane, auth: AuthState, fingerprint: str) -> str:
    payload = f"{plane}|{auth}|{fingerprint}"
    return hashlib.sha256(payload.encode("utf-8")).hexdigest()[:16]


# ---------------------------------------------------------------------------
# Cross-plane claim validation
# ---------------------------------------------------------------------------


def _is_path_cross_plane(path: str, plane: AccessPlane) -> bool:
    if plane == "web":
        return bool(_WEB_NATIVE_PATH.match(path) or _UNC.match(path))
    if plane == "api":
        return True  # API callers have no direct FS
    # terminal / app
    return bool(_URL.match(path))


# ---------------------------------------------------------------------------
# Flag evaluation
# ---------------------------------------------------------------------------


def _evaluate_flags(
    plane: AccessPlane,
    auth: AuthState,
    fs_topology: FsTopology,
    claimed_paths: List[str],
    login_time_ms: Optional[int],
) -> List[str]:
    flags: List[str] = []

    if plane == "terminal" and auth == "sudo":
        flags.append("ELEVATED_TERMINAL")
    if plane == "terminal" and auth == "service_account":
        flags.append("SERVICE_ACCOUNT_ELEVATED")
    if plane == "web" and auth == "anonymous":
        flags.append("UNAUTHENTICATED_WEB")
    if plane == "api" and auth == "anonymous":
        flags.append("UNAUTHENTICATED_API")
    if fs_topology.temp_only:
        flags.append("TEMP_FS_ONLY")
    if auth in ("authenticated", "sudo") and login_time_ms is None:
        flags.append("SESSION_MISSING_LOGIN_TIME")

    if any(_is_path_cross_plane(p, plane) for p in claimed_paths):
        flags.append("CROSS_PLANE_CLAIM")

    return flags


# ---------------------------------------------------------------------------
# Primary evaluator
# ---------------------------------------------------------------------------


def evaluate_operator_space(inp: OperatorSpaceInput) -> OperatorSpaceDecision:
    """Evaluate an operator's system-space position and return a governance decision.

    The decision integrates with GeoSeal's ring model: ``ring`` maps directly
    to the same ALLOW/QUARANTINE/ESCALATE/DENY tier thresholds used in L13.

    Examples::

        decision = evaluate_operator_space(OperatorSpaceInput(
            access_plane="terminal",
            auth_state="authenticated",
            session_fingerprint="sess-uuid-123",
            login_time_ms=1716000000000,
            claimed_paths=["/home/user/docs"],
        ))
        assert decision.ring == "core"
        assert decision.trust_score == 0.85
    """
    plane = inp.access_plane
    auth = inp.auth_state
    fingerprint = inp.session_fingerprint or f"{plane}:{auth}:anon"

    space_id = _derive_space_id(plane, auth, fingerprint)
    fs_topology = derive_fs_topology(plane, auth)
    ring, trust_score = _RING_TABLE[(plane, auth)]

    flags = _evaluate_flags(plane, auth, fs_topology, inp.claimed_paths, inp.login_time_ms)

    # Penalise trust for violations
    effective_trust = trust_score
    if "CROSS_PLANE_CLAIM" in flags:
        effective_trust = max(0.0, effective_trust - 0.30)
    if "SESSION_MISSING_LOGIN_TIME" in flags:
        effective_trust = max(0.0, effective_trust - 0.15)

    if effective_trust < 0.1:
        effective_ring: OperatorRing = "blocked"
    elif effective_trust <= 0.4:
        effective_ring = "restricted"
    else:
        effective_ring = ring  # type: ignore[assignment]

    status = f"operator_space_{plane}_{auth}"
    flag_str = f" flags=[{','.join(flags)}]" if flags else ""
    reason = f"plane={plane} auth={auth} sandbox={fs_topology.sandbox_level}{flag_str}"

    return OperatorSpaceDecision(
        ring=effective_ring,
        trust_score=round(effective_trust, 4),
        space_id=space_id,
        fs_topology=fs_topology,
        governance_flags=flags,
        status=status,
        reason=reason,
    )


# ---------------------------------------------------------------------------
# Helpers
# ---------------------------------------------------------------------------


def ring_to_governance_tier(ring: OperatorRing) -> GovernanceTier:
    """Map an OperatorRing to the L13 governance tier vocabulary."""
    mapping: dict[OperatorRing, GovernanceTier] = {
        "core": "ALLOW",
        "outer": "QUARANTINE",
        "restricted": "ESCALATE",
        "blocked": "DENY",
    }
    return mapping[ring]


def decision_to_metadata(decision: OperatorSpaceDecision) -> dict:
    """Convert a decision to a JSON-friendly plain dict (receipt format)."""
    return {
        "ring": decision.ring,
        "trust_score": decision.trust_score,
        "space_id": decision.space_id,
        "sandbox_level": decision.fs_topology.sandbox_level,
        "accessible_roots": decision.fs_topology.accessible_roots,
        "writable_paths": decision.fs_topology.writable_paths,
        "persists_across_sessions": decision.fs_topology.persists_across_sessions,
        "temp_only": decision.fs_topology.temp_only,
        "mount_points": decision.fs_topology.mount_points,
        "governance_flags": decision.governance_flags,
        "status": decision.status,
        "reason": decision.reason,
    }

File: src/geosealOperatorSpace.ts
/**
 * @file geosealOperatorSpace.ts
 * @module geosealOperatorSpace
 * @layer Layer 3, Layer 13
 *
 * GeoSeal Operator System-Space Model
 *
 * Extends GeoSeal's geographic ring model to cover the operator's position in
 * **system topology space** — not physical coordinates, but the structural
 * location of the operator: which access plane they arrived through (web /
 * terminal / app / API) and whether they are authenticated.
 *
 * The access plane + auth state together determine:
 *   - A file-system topology (what paths are reachable and writable)
 *   - A governance ring (core / outer / restricted / blocked)
 *   - A deterministic space ID analogous to the geographic geoid
 *   - Governance flags that surface cross-plane claims and auth anomalies
 *
 * Why this matters:
 *   A web-anonymous operator claiming access to /home/user/documents is
 *   structurally impossible — their sandbox grants no native FS at all.
 *   The governance layer should catch the claim before it reaches the FS.
 *   Similarly, a terminal+sudo operator carries elevated risk that warrants
 *   its own flag even at high trust, so downstream layers can gate writes.
 */

import * as crypto from 'node:crypto';

// ─────────────────────────────────────────────────────────────────────────────
// Enumerations
// ─────────────────────────────────────────────────────────────────────────────

/**
 * The structural plane through which an operator connects to the system.
 * Each plane has a fixed, deterministic FS topology regardless of auth state.
 */
export type OperatorAccessPlane = 'web' | 'terminal' | 'app' | 'api';

/**
 * Authentication status of the operator at session time.
 * Drives trust score and writable-path access within each plane.
 */
export type OperatorAuthState = 'authenticated' | 'anonymous' | 'service_account' | 'sudo';

/** Degree to which the operator's file-system access is sandboxed. */
export type SandboxLevel = 'none' | 'partial' | 'full';

/** GeoSeal governance ring — mirrors the geographic ring vocabulary. */
export type OperatorRing = 'core' | 'outer' | 'restricted' | 'blocked';

// ─────────────────────────────────────────────────────────────────────────────
// FS Topology
// ─────────────────────────────────────────────────────────────────────────────

/**
 * File-system access topology derived from the operator's system-space position.
 * Describes structural constraints, not a list of guaranteed paths.
 */
export interface FsTopology {
  /** How tightly sandboxed is this operator's FS access? */
  sandboxLevel: SandboxLevel;
  /**
   * Root paths reachable in principle for this plane+auth combination.
   * Empty for full-sandbox planes (browser context).
   * These are canonical examples — real enforcement is the caller's job.
   */
  accessibleRoots: string[];
  /** Subset of accessibleRoots where writes are permitted by default. */
  writablePaths: string[];
  /** Whether FS state persists after the session ends. */
  persistsAcrossSessions: boolean;
  /**
   * True when the operator can only write to a temporary/ephemeral location
   * (e.g. /tmp or browser OPFS without explicit permission grant).
   */
  tempOnly: boolean;
  /**
   * OS-level mount points visible to this operator.
   * Empty for browser and anonymous-API planes.
   */
  mountPoints: string[];
}

// ─────────────────────────────────────────────────────────────────────────────
// Input / Output types
// ─────────────────────────────────────────────────────────────────────────────

export interface OperatorSpaceInput {
  /** How the operator is connected. */
  accessPlane: OperatorAccessPlane;
  /** Auth state at session time. */
  authState: OperatorAuthState;
  /**
   * Opaque session fingerprint for space-ID derivation (e.g. session cookie
   * hash, terminal session UUID, app install ID). Must NOT contain PII.
   * If omitted, an anonymous fingerprint is derived from plane+auth only.
   */
  sessionFingerprint?: string;
  /** Unix epoch ms when the session was authenticated. Null for anonymous. */
  loginTimeMs?: number;
  /**
   * Paths the operator is claiming access to.
   * Used for cross-plane claim validation.
   */
  claimedPaths?: string[];
}

/** Governance flags surfaced by the space evaluation. */
export type GovernanceFlag =
  | 'ELEVATED_TERMINAL' // terminal + sudo
  | 'UNAUTHENTICATED_WEB' // web + anonymous (highest risk profile)
  | 'UNAUTHENTICATED_API' // api + anonymous
  | 'CROSS_PLANE_CLAIM' // claimed paths inconsistent with access plane
  | 'SESSION_MISSING_LOGIN_TIME' // auth says authenticated but no loginTimeMs
  | 'TEMP_FS_ONLY' // full sandbox, all writes are ephemeral
  | 'SERVICE_ACCOUNT_ELEVATED'; // service_account on terminal plane

export interface OperatorSpaceDecision {
  /** Canonical ring for downstream governance gates. */
  ring: OperatorRing;
  /** Composite trust score [0.0, 1.0]. Higher = more trusted. */
  trustScore: number;
  /**
   * Deterministic identifier for this operator's system-space position.
   * Analogous to the geographic geoid — stable across calls with the same
   * plane+auth+fingerprint, usable as a receipt anchor.
   */
  spaceId: string;
  /** Derived FS topology for this plane + auth combination. */
  fsTopology: FsTopology;
  /** Governance flags raised during evaluation. */
  governanceFlags: GovernanceFlag[];
  /** Machine-readable status tag. */
  status: string;
  /** Human-readable explanation. */
  reason: string;
}

// ─────────────────────────────────────────────────────────────────────────────
// FS topology derivation
// ─────────────────────────────────────────────────────────────────────────────

/**
 * Returns the structural FS topology for a given plane + auth combination.
 * These are deterministic governance constraints, not live filesystem probes.
 */
export function derivefsTopology(plane: OperatorAccessPlane, auth: OperatorAuthState): FsTopology {
  switch (plane) {
    case 'web':
      // Browser context — OPFS only (sandboxed, origin-scoped, no native FS)
      return {
        sandboxLevel: 'full',
        accessibleRoots: [], // no native root access
        writablePaths: [], // OPFS is not representable as a native path
        persistsAcrossSessions: auth === 'authenticated', // OPFS persists when logged in
        tempOnly: auth === 'anonymous',
        mountPoints: [], // no mount points visible in browser
      };

    case 'terminal':
      if (auth === 'sudo') {
        return {
          sandboxLevel: 'none',
          accessibleRoots: ['/', 'C:\\'],
          writablePaths: ['/', 'C:\\'],
          persistsAcrossSessions: true,
          tempOnly: false,
          mountPoints: ['/', '/proc', '/sys', '/dev', 'C:\\', 'D:\\'],
        };
      }
      if (auth === 'authenticated' || auth === 'service_account') {
        return {
          sandboxLevel: 'none',
          accessibleRoots: ['~', '/home', '/var', '/tmp', 'C:\\Users', 'C:\\ProgramData'],
          writablePaths: ['~', '/tmp', 'C:\\Users\\{user}', 'C:\\Temp'],
          persistsAcrossSessions: true,
          tempOnly: false,
          mountPoints: ['/', 'C:\\', 'D:\\'],
        };
      }
      // anonymous terminal — /tmp only, no persistent FS
      return {
        sandboxLevel: 'partial',
        accessibleRoots: ['/tmp', 'C:\\Temp'],
        writablePaths: ['/tmp', 'C:\\Temp'],
        persistsAcrossSessions: false,
        tempOnly: true,
        mountPoints: [],
      };

    case 'app':
      if (auth === 'anonymous') {
        return {
          sandboxLevel: 'partial',
          accessibleRoots: [],
          writablePaths: [],
          persistsAcrossSessions: false,
          tempOnly: true,
          mountPoints: [],
        };
      }
      // app sandbox — app-data directory only (no system paths)
      return {
        sandboxLevel: 'partial',
        accessibleRoots: ['~/AppData', '~/.local/share', '~/Library/Application Support'],
        writablePaths: ['~/AppData/{app}', '~/.local/share/{app}'],
        persistsAcrossSessions: true,
        tempOnly: false,
        mountPoints: [],
      };

    case 'api':
      // API callers have no direct FS — server-side FS only, mediated by the service
      return {
        sandboxLevel: auth === 'anonymous' ? 'full' : 'partial',
        accessibleRoots: [],
        writablePaths: [],
        persistsAcrossSessions: false,
        tempOnly: true,
        mountPoints: [],
      };
  }
}

// ─────────────────────────────────────────────────────────────────────────────
// Ring + trust derivation
// ─────────────────────────────────────────────────────────────────────────────

interface RingProfile {
  ring: OperatorRing;
  trustScore: number;
}

/** Lookup table: plane → auth → ring + trust */
const RING_TABLE: Record<OperatorAccessPlane, Record<OperatorAuthState, RingProfile>> = {
  terminal: {
    sudo: { ring: 'core', trustScore: 0.95 },
    authenticated: { ring: 'core', trustScore: 0.85 },
    service_account: { ring: 'core', trustScore: 0.8 },
    anonymous: { ring: 'restricted', trustScore: 0.35 },
  },
  app: {
    authenticated: { ring: 'outer', trustScore: 0.7 },
    service_account: { ring: 'outer', trustScore: 0.65 },
    anonymous: { ring: 'restricted', trustScore: 0.25 },
    sudo: { ring: 'outer', trustScore: 0.7 }, // apps don't have sudo; treat as authenticated
  },
  api: {
    service_account: { ring: 'core', trustScore: 0.8 },
    authenticated: { ring: 'outer', trustScore: 0.7 },
    anonymous: { ring: 'blocked', trustScore: 0.05 },
    sudo: { ring: 'outer', trustScore: 0.7 }, // API + sudo treated as authenticated
  },
  web: {
    authenticated: { ring: 'outer', trustScore: 0.6 },
    anonymous: { ring: 'blocked', trustScore: 0.05 },
    service_account: { ring: 'outer', trustScore: 0.55 },
    sudo: { ring: 'outer', trustScore: 0.6 }, // web + sudo treated as authenticated
  },
};

// ─────────────────────────────────────────────────────────────────────────────
// Space ID
// ─────────────────────────────────────────────────────────────────────────────

/**
 * Deterministic space identifier for this operator's system-space position.
 * Analogous to the geographic geoid — stable across calls with the same inputs,
 * usable as a receipt anchor in audit logs.
 */
function deriveSpaceId(
  plane: OperatorAccessPlane,
  auth: OperatorAuthState,
  fingerprint: string
): string {
  const payload = `${plane}|${auth}|${fingerprint}`;
  return crypto.createHash('sha256').update(payload, 'utf8').digest('hex').slice(0, 16);
}

// ─────────────────────────────────────────────────────────────────────────────
// Cross-plane claim validation
// ─────────────────────────────────────────────────────────────────────────────

const WEB_NATIVE_PATH_PATTERN = /^[/\\]|^[A-Za-z]:\\/;
const URL_PATTERN = /^https?:\/\//i;
const UNC_PATTERN = /^\\\\/;

/**
 * Returns true if the claimed path is structurally inconsistent with the
 * operator's access plane.
 *
 * - Web operators cannot claim native FS paths.
 * - API-anonymous operators cannot claim any paths.
 * - Terminal operators cannot claim bare HTTP URLs as file paths.
 */
function isPathCrossPlane(path: string, plane: OperatorAccessPlane): boolean {
  switch (plane) {
    case 'web':
      // Browser has no native FS — claiming any local or UNC path is a violation
      return WEB_NATIVE_PATH_PATTERN.test(path) || UNC_PATTERN.test(path);
    case 'api':
      // API callers have no direct FS at all
      return true;
    case 'terminal':
    case 'app':
      // Terminal/app operators should not be claiming bare HTTP URLs as file paths
      return URL_PATTERN.test(path);
    default:
      return false;
  }
}

// ─────────────────────────────────────────────────────────────────────────────
// Governance flag evaluation
// ─────────────────────────────────────────────────────────────────────────────

function evaluateFlags(
  plane: OperatorAccessPlane,
  auth: OperatorAuthState,
  fsTopology: FsTopology,
  claimedPaths: string[],
  loginTimeMs: number | undefined
): GovernanceFlag[] {
  const flags: GovernanceFlag[] = [];

  if (plane === 'terminal' && auth === 'sudo') flags.push('ELEVATED_TERMINAL');
  if (plane === 'terminal' && auth === 'service_account') flags.push('SERVICE_ACCOUNT_ELEVATED');
  if (plane === 'web' && auth === 'anonymous') flags.push('UNAUTHENTICATED_WEB');
  if (plane === 'api' && auth === 'anonymous') flags.push('UNAUTHENTICATED_API');
  if (fsTopology.tempOnly) flags.push('TEMP_FS_ONLY');

  if ((auth === 'authenticated' || auth === 'sudo') && loginTimeMs == null) {
    flags.push('SESSION_MISSING_LOGIN_TIME');
  }

  const crossPlanePaths = claimedPaths.filter((p) => isPathCrossPlane(p, plane));
  if (crossPlanePaths.length > 0) flags.push('CROSS_PLANE_CLAIM');

  return flags;
}

// ─────────────────────────────────────────────────────────────────────────────
// Primary evaluation function
// ─────────────────────────────────────────────────────────────────────────────

/**
 * Evaluate an operator's system-space position and return a governance decision.
 *
 * The decision integrates with GeoSeal's ring model: `ring` maps directly to
 * the same ALLOW/QUARANTINE/ESCALATE/DENY tier thresholds used in L13.
 *
 * @example
 * ```ts
 * const decision = evaluateOperatorSpace({
 *   accessPlane: 'terminal',
 *   authState: 'authenticated',
 *   sessionFingerprint: sessionUUID,
 *   loginTimeMs: Date.now(),
 *   claimedPaths: ['/home/user/docs'],
 * });
 * // decision.ring === 'core', decision.trustScore === 0.85
 * ```
 */
export function evaluateOperatorSpace(input: OperatorSpaceInput): OperatorSpaceDecision {
  const { accessPlane, authState, sessionFingerprint, loginTimeMs, claimedPaths = [] } = input;

  const fingerprint =
    sessionFingerprint && sessionFingerprint.length > 0
      ? sessionFingerprint
      : `${accessPlane}:${authState}:anon`;

  const spaceId = deriveSpaceId(accessPlane, authState, fingerprint);
  const fsTopology = derivefsTopology(accessPlane, authState);
  const { ring, trustScore } = RING_TABLE[accessPlane][authState];

  const flags = evaluateFlags(accessPlane, authState, fsTopology, claimedPaths, loginTimeMs);

  // Penalize trust score for governance violations
  let effectiveTrust = trustScore;
  if (flags.includes('CROSS_PLANE_CLAIM')) effectiveTrust = Math.max(0, effectiveTrust - 0.3);
  if (flags.includes('SESSION_MISSING_LOGIN_TIME'))
    effectiveTrust = Math.max(0, effectiveTrust - 0.15);

  const effectiveRing: OperatorRing =
    effectiveTrust < 0.1 ? 'blocked' : effectiveTrust <= 0.4 ? 'restricted' : ring;

  const status = `operator_space_${accessPlane}_${authState}`;
  const reason =
    `plane=${accessPlane} auth=${authState} sandbox=${fsTopology.sandboxLevel}` +
    (flags.length > 0 ? ` flags=[${flags.join(',')}]` : '');

  return {
    ring: effectiveRing,
    trustScore: Math.round(effectiveTrust * 10000) / 10000,
    spaceId,
    fsTopology,
    governanceFlags: flags,
    status,
    reason,
  };
}

// ─────────────────────────────────────────────────────────────────────────────
// Serialisation helpers
// ─────────────────────────────────────────────────────────────────────────────

/** Convert an OperatorSpaceDecision to a JSON-friendly plain object. */
export function operatorSpaceToRecord(decision: OperatorSpaceDecision): Record<string, unknown> {
  return {
    ring: decision.ring,
    trust_score: decision.trustScore,
    space_id: decision.spaceId,
    sandbox_level: decision.fsTopology.sandboxLevel,
    accessible_roots: decision.fsTopology.accessibleRoots,
    writable_paths: decision.fsTopology.writablePaths,
    persists_across_sessions: decision.fsTopology.persistsAcrossSessions,
    temp_only: decision.fsTopology.tempOnly,
    mount_points: decision.fsTopology.mountPoints,
    governance_flags: decision.governanceFlags,
    status: decision.status,
    reason: decision.reason,
  };
}

/**
 * Maps an OperatorRing to the L13 governance tier vocabulary.
 * Integrates operator-space decisions into the main 14-layer pipeline.
 */
export function ringToGovernanceTier(
  ring: OperatorRing
): 'ALLOW' | 'QUARANTINE' | 'ESCALATE' | 'DENY' {
  switch (ring) {
    case 'core':
      return 'ALLOW';
    case 'outer':
      return 'QUARANTINE';
    case 'restricted':
      return 'ESCALATE';
    case 'blocked':
      return 'DENY';
  }
}
